x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportOpenJobAction.java

```diff
@@ -438,7 +443,7 @@ public OpenJobPersistentTasksExecutor(Settings settings, ClusterService clusterS
             JobNodeSelector jobNodeSelector = new JobNodeSelector(clusterState, jobId, MlTasks.JOB_TASK_NAME, memoryTracker,
                 job.allowLazyOpen() ? Integer.MAX_VALUE : maxLazyMLNodes, node -> nodeFilter(node, job));
             return jobNodeSelector.selectNode(
-                maxOpenJobs, maxConcurrentJobAllocations, maxMachineMemoryPercent, isMemoryTrackerRecentlyRefreshed);
+                maxOpenJobs, maxConcurrentJobAllocations, maxMachineMemoryPercent, maxJobMemoryBytes, isMemoryTrackerRecentlyRefreshed);
```

Comment:
The check of 
>  if autoscaling is being used 

Is tricky. I wonder if it is as simple as to check if the plugin is enabled or not.

@henningandersen what do you think?

Reply:
I would prefer to decouple the two concerns. We should assume autoscaling is always "enabled". It now (or rather after #61575) returns a capacity that could be useful for alerting on-prem too and while the actual scaling in that case may not be automatic, the detection of a need to scale could be.

I am no expert on the ML code, but AFAICS, if we pass down `Long.MAX_VALUE` when `max_model_memory_limit` is unset, the code in `selectNode` should behave as it did prior to this change?

On this specific part:
> if we think the user might add more nodes, so we'll have to assume they won't.
I think you meant "larger nodes"? It is my understanding that hitting the limit for the recommended 64GB ML production nodes is unexpected. If a user started smaller (which could be fine), they might simply want to use bigger nodes even on-prem (which could be in a cloud) when the job cannot run?

About 2:
> Make it an error for autoscaling to be used if max_model_memory_limit is unlimited.
I think this is coupling the concerns too tightly. In our cloud, things would be configured correctly. Other users of the autoscaling API (like on-prem alerting) would work fine, with the reaction resulting in either setting the limit or upgrading the nodes?

About 4:
> If autoscaling is not being used then ...
Would this not have to use the `max_model_memory_limit` if set and only if not revert to `max_machine_memory_percent * ml.machine_memory`? I fail to see how this would also not work out fine in both cloud and on-prem?

My response to 2 and 4 are clearly contradictory :slightly_smiling_face: , I think either way works fine with my autoscaling glasses on.

Reply:
The way things work at the moment, if you try to open an ML job that has a model memory requirement of `X` and no node in the cluster has space for an ML job of size `X` then unless that specific job config says it's allowed to open lazily the open request will fail.

One of the changes in this PR is trying to move from "no node in the cluster has space for an ML job of size `X`" to "no node in the cluster or that might be added to the cluster has space for an ML job of size `X`".

> I think you meant "larger nodes"?

It could be larger nodes or more nodes.  Larger nodes in the case where somebody is trying to open a job that wouldn't fit on any current ML node even if no other jobs were running.  But there is also the case where somebody is trying to open a job that would fit on an existing ML node if it was empty, but will not fit anywhere due to already-running ML jobs.

> It is my understanding that hitting the limit for the recommended 64GB ML production nodes is unexpected.

This is related - it's unexpected that somebody would open a single job that wouldn't fit on an empty 64GB ML node.  But if they already have 10 jobs running on that node then quite a small job might not fit on it.  Also, although we have a best practices doc that says it's best to run 64GB ML nodes in production I know there are quite big ML users who don't run such big nodes in their self-managed clusters.

> Would this not have to use the `max_model_memory_limit` if set and only if not revert to `max_machine_memory_percent * ml.machine_memory`? I fail to see how this would also not work out fine in both cloud and on-prem?

There are edge cases outside of Cloud.  In Cloud the intention is that `max_model_memory_limit` is set such that jobs cannot be created with a model memory limit that would be too big to ever run.  In self-managed it is possible that this setting is used for another reason, for example an administrator might want to prevent users creating single jobs that can use all their ML capacity.  The setting only applies when the model memory limit configuration is changed.  So the following scenario is possible:

- Cluster has a single 8GB ML node
- Cluster starts off with `max_model_memory_limit` unset
- User creates a job with a model memory limit of 2GB
- `max_model_memory_limit` is set to 1GB
- User opens their job with a model memory limit of 2GB and it runs fine

But this scenario wouldn't work if we are using `max_model_memory_limit` as a proxy for "what is the biggest job that could ever be run in this cluster".

All these problems come from trying to avoid having a new setting for "what is the biggest ML node that could ever be added to this cluster".  Maybe it's best to just introduce that setting, have Cloud set it, and if it's unset make the default "the biggest ML node currently in the cluster".  I guess it could vary between AWS, GCP and Azure anyway, as the nominal 64GB node might be 58GB or 62GB in some regions.

Reply:
I think we should stick to decoupling this as originally planned. 

Adding a new setting is OK, IMO. Its lower overhead and by default it is current maximum node size in the cluster.